### DISCUSSION TITLE
bn table output

### DISCUSSION THREAD

**Discussion by kiminh** [Qihoo360/tensornet#67] [General] [answered]
there has bn table output in the model file accompanied by sparse_table and dense_table.  @gzm55 @jiangxinglei @thundercatli 

my question is that
1. is the bn table output included in the tf graph?
2. how to serve the model file with bn table? should I implement the customized logic of batch normalization in cpp or java. If convenient, would you please provide a serving demo with bn table.

**Comment by thundercatli** [comment]
bn related variables moving_mean and meaning_var are included in tf checkpoint and can be used for predict directly. BN output file stores summary statistics like total number, total sum and total square sum for backend calculation.

**Reply by kiminh** [reply]
> bn related variables moving_mean and meaning_var are included in tf checkpoint and can be used for predict directly. BN output file stores summary statistics like total number, total sum and total square sum for backend calculation.

what do you mean for backend calculation?  it will be used for incremental training?

**Reply by thundercatli** [reply]
Yes, statistics will be loaded for next training lifecycle. BN layer will collect data from all nodes, it has a max_count option to limit total records, if exceeds, will do moving.

**Reply by kiminh** [reply]
> Yes, statistics will be loaded for next training lifecycle. BN layer will collect data from all nodes, it has a max_count option to limit total records, if exceeds, will do moving.

thanks for your reply 

![image](https://github.com/user-attachments/assets/b6856581-ea59-492b-9e0a-b7ce79f0a339)
![image](https://github.com/user-attachments/assets/820602f7-23a2-448a-8e83-28db7c052f00)


1. I dived into the code.
https://github.com/Qihoo360/tensornet/blob/master/examples/models/wide_deep.py#L52

the max_count is set to 100000 default. I mean in industrial setting, the input before the main net is equal to 
```
the length of continuous feature + (number of single-value slots) * embedding_dim + (number of multi-value slots) * embedding_dim
```
so in general, it's totally enough?

2. I mean bn can always be added in practice, in the official model implementation, it's only used in wide and deep.  is that not suitable for dcn or deepfm?
3. what's the difference of PCTRDNNBatchNormalization vs TNBatchNormalizationBase. is there any references?

**Comment by thundercatli** [comment]
1. max_count is used to limit number of input samples, not features
2. wide_deep example is used for ut or debug. does't mean bn is not suitable for other network. 
3. check the comments. just algorithm diff

**Reply by kiminh** [reply]
> 1. max_count is used to limit number of input samples, not features
> 2. wide_deep example is used for ut or debug. does't mean bn is not suitable for other network.
> 3. check the comments. just algorithm diff

1. if I have 1 billion training samples, max_count should be set to 1 billion?

**Reply by thundercatli** [reply, marked as answer]
It depends. If you think 1000 records are sufficient to represent your dataset, it can be set to 1000. Or you want a larger number, may be 1%/5%/10% of your total training set, it also works. No need to set a number to cover all samples. It's a Hyperparameter

**Reply by kiminh** [reply]
> It depends. If you think 1000 records are sufficient to represent your dataset, it can be set to 1000. Or you want a larger number, may be 1%/5%/10% of your total training set, it also works. No need to set a number to cover all samples. It's a Hyperparameter

got it, thanks.
